### PR TITLE
[clang] fix divide by zero in ComplexExprEvaluator

### LINKED ISSUE #55390: ComplexExprEvaluator - Assertion failed: RHS.U.VAL != 0 && "Divide by zero?"
Reduced from https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=36067
```
auto f() { return 43273096 / 65536j; }
```
```
C:\LLVM>stage1\bin\clang -c -S -O3 fuzz.cpp -o -
Assertion failed: RHS.U.VAL != 0 && "Divide by zero?", file C:\LLVM\llvm-project\llvm\lib\Support\APInt.cpp, line 1585
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: stage1\\bin\\clang -c -S -O3 fuzz.cpp -o -
1.      fuzz.cpp:1:33: current parser token ';'
2.      fuzz.cpp:1:9: parsing function body 'f'
3.      fuzz.cpp:1:9: in compound statement ('{}')
 #0 0x00007ff6c173e885 HandleAbort C:\LLVM\llvm-project\llvm\lib\Support\Windows\Signals.inc:418:0
 #1 0x00007ff86a26cd84 (C:\Windows\System32\ucrtbase.dll+0x7cd84)
 #2 0x00007ff86a26dd61 (C:\Windows\System32\ucrtbase.dll+0x7dd61)
 #3 0x00007ff86a26f7ea (C:\Windows\System32\ucrtbase.dll+0x7f7ea)
 #4 0x00007ff86a26f6e1 (C:\Windows\System32\ucrtbase.dll+0x7f6e1)
 #5 0x00007ff86a26f981 (C:\Windows\System32\ucrtbase.dll+0x7f981)
 #6 0x00007ff6c175c34a llvm::APInt::udiv(class llvm::APInt const &) const C:\LLVM\llvm-project\llvm\lib\Support\APInt.cpp:1586:0
 #7 0x00007ff6c1757de3 llvm::APInt::sdiv(class llvm::APInt const &) const C:\LLVM\llvm-project\llvm\lib\Support\APInt.cpp:1660:0
 #8 0x00007ff6c4bd87f9 llvm::APSInt::operator/(class llvm::APSInt const &) const C:\LLVM\llvm-project\llvm\include\llvm\ADT\APSInt.h:134:0
 #9 0x00007ff6c4c0981a `anonymous namespace'::ComplexExprEvaluator::VisitBinaryOperator C:\LLVM\llvm-project\clang\lib\AST\ExprConstant.cpp:14468:0
```

### PR DESCRIPTION
**Pull request by c8ef**
fix: #55390.

### CODE REVIEW

File: clang/test/SemaCXX/complex-div.cpp [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// RUN: %clang_cc1 -fsyntax-only %s
+
+auto f() { return 43273096 / 65536j; }
```
Comment (cor3ntin) [inline clang/test/SemaCXX/complex-div.cpp:3]: can you move that to , for example `clang/test/SemaCXX/constant-expression-cxx11.cpp`
and make sure that actually always get evaluated?
maybe `constexpr auto a = `43273096 / 65536j;`

**Comment by cor3ntin** [comment]
This change needs a release note.
Please add an entry to `clang/docs/ReleaseNotes.rst` in the section the most adapted to the change, and referencing any Github issue this change fixes. Thanks!

The fix itself looks good

**Review by cor3ntin — APPROVED** [review]

**Comment by cor3ntin** [comment]
@c8ef you need me to merge that for you?

File: clang/lib/AST/ExprConstant.cpp [hunk 2]
```diff
@@ -15567,12 +15567,13 @@ bool ComplexExprEvaluator::VisitBinaryOperator(const BinaryOperator *E) {
...
       }
     } else {
-      if (RHS.getComplexIntReal() == 0 && RHS.getComplexIntImag() == 0)
-        return Error(E, diag::note_expr_divide_by_zero);
-
       ComplexValue LHS = Result;
       APSInt Den = RHS.getComplexIntReal() * RHS.getComplexIntReal() +
         RHS.getComplexIntImag() * RHS.getComplexIntImag();
+      if ((RHS.getComplexIntReal() == 0 && RHS.getComplexIntImag() == 0) ||
+          Den.isZero())
```
Comment (s-barannikov) [inline clang/lib/AST/ExprConstant.cpp:15574]: (nit) The first part of the check implies the second, why not just
```suggestion
      if (Den.isZero())
```

**Review by s-barannikov — APPROVED** [review]

**Comment by efriedma-quic** [comment]
Opened #104738 as a followup.

**Comment by shafik** [comment]
For future reference the summary here is not really sufficient. Since this what usually goes in the git log they should be enough details for folks reading the git log to understand what the change was.

The summary should explain the problem and provide a summary of the fix as well.

**Pull request merged**